### PR TITLE
Trace static level filtering

### PR DESCRIPTION
**Pull request by phungleson**
## Motivation

`tokio-trace` should have static verbosity level filtering, like the `log` crate. The static max verbosity level should be controlled at compile time with a set of features. It should be possible to set a separate max level for release and debug mode builds.

https://github.com/tokio-rs/tokio/issues/959

## Solution

We can do this fairly similarly to how the `log` crate does it: `tokio-trace` should export a constant whose value is set based on the static max level feature flags. Then, we can add an if statement to the `span!` and `event!` macros, like

### CODE REVIEW

File: tokio-trace/src/lib.rs [hunk 1]
```diff
@@ -345,3 +347,58 @@ pub mod subscriber;
 mod sealed {
     pub trait Sealed {}
 }
+
+/// An enum representing the available verbosity level filters of the logger.
+#[repr(usize)]
+#[derive(Copy, Clone, Debug, Hash)]
+pub enum LevelFilter {
```
Comment (hawkw) [inline tokio-trace/src/lib.rs]: For forwards-compatibility reasons, we try to avoid having enum types in public APIs. Can this be a struct wrapping an inner private enum type instead? Refer to how the `Level` type in `tokio-trace-core` is implemented:
https://github.com/tokio-rs/tokio/blob/92d51202ef5e6b99004d77300de1aba0f8835513/tokio-trace/tokio-trace-core/src/metadata.rs#L224-L272

File: tokio-trace/src/lib.rs [hunk 2]
```diff
@@ -309,10 +309,12 @@
 //! [`tokio-trace-futures`]: https://github.com/tokio-rs/tokio-trace-nursery/tree/master/tokio-trace-futures
 //! [`tokio-trace-fmt`]: https://github.com/tokio-rs/tokio-trace-nursery/tree/master/tokio-trace-fmt
 //! [`tokio-trace-log`]: https://github.com/tokio-rs/tokio-trace-nursery/tree/master/tokio-trace-log
+extern crate cfg_if;
 extern crate tokio_trace_core;
 
 // Somehow this `use` statement is necessary for us to re-export the `core`
 // macros on Rust 1.26.0. I'm not sure how this makes it work, but it does.
+use cfg_if::cfg_if;
```
Comment (hawkw) [inline tokio-trace/src/lib.rs]: Importing macros this way (with the `use` keyword) doesn't work on Rust 1.26, `tokio`'s minimum supported Rust version (this is why your CI build is failing).

We'll need to import the macro the "old-fashioned" way instead, by putting the `#[macro_use]` attribute on the `extern crate` statement.

File: tokio-trace/src/lib.rs [hunk 3]
```diff
@@ -345,3 +347,58 @@ pub mod subscriber;
...
+/// The statically resolved maximum trace level.
+///
+/// See the crate level documentation for information on how to configure this.
+///
+/// This value is checked by the `event` macro. Code that manually calls functions on that value
+/// should compare the level against this value.
+///
+pub const STATIC_MAX_LEVEL: LevelFilter = MAX_LEVEL_INNER;
+
+cfg_if! {
```
Comment (hawkw) [inline tokio-trace/src/lib.rs]: Although I'm willing to add a dependency on `cfg_if`, I'm not _convinced_ we need to --- can this not be done simply using `#[cfg]` attributes?

File: tokio-trace/src/lib.rs (hunk 1 above)
Comment (phungleson) [reply tokio-trace/src/lib.rs]: @hawkw Thanks for this, if we already have this `Level`, perhaps we can just use `Level` for this purpose?

Only `Off` level is missing, do we actually need it? what do you think?

**Comment by phungleson** [comment]
@hawkw Hey mate, thanks again I have updated the code to the best I can understand (read: not much about `#cfg` 😂). Please give some thoughts about the implementation.

If it is ok (or almost there) I can start looking into tests for this feature.

File: tokio-trace/src/lib.rs (hunk 1 above)
Comment (hawkw) [reply tokio-trace/src/lib.rs]: > @hawkw Thanks for this, if we already have this `Level`, perhaps we can just use `Level` for this purpose?
> 
> Only `Off` level is missing, do we actually need it? what do you think?

I think we _will_ want an `OFF` level for the static filter, so that `tokio-trace` can be completely disabled in release mode, etc. So, we'll probably need a separate `LevelFilter` type. 

The implementation can be the same as `tokio-trace-core`s `Level` type, but with the addition of an `OFF` level. We will want to implement the [`PartialOrd`](https://doc.rust-lang.org/std/cmp/trait.PartialOrd.html) trait for comparing `LevelFilter`s with `Level`s, (`impl PartialOrd<Level> for LevelFilter`, with the same ordering as for `Level`s, but with `LevelFilter::OFF` considered less than all `Level`s

> It seems to be private struct though.

The `Level` struct is public:
https://github.com/tokio-rs/tokio/blob/92d51202ef5e6b99004d77300de1aba0f8835513/tokio-trace/tokio-trace-core/src/metadata.rs#L129
while the `LevelInner` enum is private:
https://github.com/tokio-rs/tokio/blob/92d51202ef5e6b99004d77300de1aba0f8835513/tokio-trace/tokio-trace-core/src/metadata.rs#L251

We'll want our `LevelFilter` type to follow the same pattern.

File: tokio-trace/src/macros.rs [hunk 4]
```diff
@@ -101,7 +101,7 @@ macro_rules! span {
         $name:expr,
         $($k:ident $( = $val:expr )* ),*
     ) => {
-        {
+        if $lvl <= $crate::STATIC_MAX_LEVEL {
```
Comment (hawkw) [inline tokio-trace/src/macros.rs]: 👍 this seems correct to me!

**Comment by phungleson** [comment]
Hey @hawkw I have updated the implementation, can you check again? cheers,

**Comment by hawkw** [comment]
It looks like the [CI build failure](https://dev.azure.com/tokio-rs/Tokio/_build/results?buildId=175) is due to an error installing Rust, so that's not your fault. I'm going to restart the build.

File: tokio-trace/src/level_filters.rs [hunk 5]
```diff
@@ -0,0 +1,121 @@
+use std::cmp::Ordering;
+use tokio_trace_core::Level;
+
+/// Describes the level of verbosity of a span or event.
+#[derive(Clone, Debug, Eq, PartialEq, Ord, PartialOrd)]
+pub struct LevelFilter(LevelFilterInner, Option<Level>);
```
Comment (hawkw) [inline tokio-trace/src/level_filters.rs]: Adding an `Option<Level>` to the `LevelFilter` type is an interesting approach I hadn't actually thought of. I would have just had a separate `LevelFilterInner` type and implemented the comparison with `Level` via a match.  Your solution seems more elegant to me --- good thinking!

With that said, I think the `LevelFilterInner` type shouldn't actually be necessary --- a `LevelFilter` should be representable as just
```suggestion
pub struct LevelFilter(Option<Level>);
```

File: tokio-trace/src/level_filters.rs [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+
+impl LevelFilter {
+    /// The "zero" level.
+    ///
+    /// Designates no logging.
+    pub const ZERO: LevelFilter = LevelFilter(LevelFilterInner::Zero, None);
```
Comment (hawkw) [inline tokio-trace/src/level_filters.rs]: Minor naming nit, but: can we call this `OFF` instead of `ZERO`?
```suggestion
    pub const OFF: LevelFilter = LevelFilter(LevelFilterInner::Zero, None);
```

File: tokio-trace/src/level_filters.rs [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+    ///
+    /// Designates lower priority information.
+    Debug,
+    /// The "trace" level.
+    ///
+    /// Designates very low priority, often extremely verbose, information.
+    Trace,
+}
+
+impl PartialEq<LevelFilter> for Level {
```
Comment (hawkw) [inline tokio-trace/src/level_filters.rs]: I think the reason these traits had to be implemented for `Level` rather than for `LevelFilter` is because the `Level` is on the right-hand side of the comparison and the `LevelFilter` is on the left. We could reverse the comparison if we wanted the impl to be `PartialEq<Level> for LevelFilter` instead, like this:
```rust
if $crate::level_filters::STATIC_MAX_LEVEL > $lvl
```

However, this seems to work, so I'm fine either way.

Comment (hawkw) [reply tokio-trace/src/level_filters.rs]: As a side note, I'm actually surprised the compiler accepts this due to the [orphan rules](http://smallcultfollowing.com/babysteps/blog/2015/01/14/little-orphan-impls/); neither `PartialEq` nor `Level` are defined in this crate. I suppose since the `PartialEq` (and `PartialOrd`) traits are being implemented with a _type parameter_ defined in this crate (`LevelFilter`) this is considered acceptable.

I've just never seen this before --- it's possible the compiler only recently started accepting this?

File: tokio-trace/src/level_filters.rs [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+
+/// The statically resolved maximum trace level.
+///
+/// See the crate level documentation for information on how to configure this.
+///
+/// This value is checked by the `event` macro. Code that manually calls functions on that value
+/// should compare the level against this value.
+pub const STATIC_MAX_LEVEL: LevelFilter = LevelFilter::ZERO;
+
+#[cfg(feature = "max_level_zero")]
```
Comment (hawkw) [inline tokio-trace/src/level_filters.rs]: I just noticed this, but I think these `cfg` attributes might have to be a bit more complex --- since we define a different value for `STATIC_MAX_LEVEL` in release mode if there is a release max level, the cfg attributes for the `max_level_XXX` features will actually have to be
```rust
#[cfg(all(
    feature = "max_level_XXX", 
    not(all(
        any(
             feature = "release_max_level_zero",  feature = "release_max_level_error", 
             feature = "release_max_level_warn", ...
        ),
       debug_assertions
    )
)]
```
I think the reason that the `log` crate uses the `cfg_if` macro is probably to avoid this complexity, and given that it's quite ugly, I'd be happy to switch back to using `cfg_if` (despite my suggestion from earlier).

Sorry for giving incorrect advice! :)

File: tokio-trace/src/level_filters.rs [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+use tokio_trace_core::Level;
+
+/// Describes the level of verbosity of a span or event.
+#[derive(Clone, Debug, Eq, PartialEq, Ord, PartialOrd)]
+pub struct LevelFilter(Option<Level>);
+
+impl LevelFilter {
+    /// The "off" level.
+    ///
+    /// Designates to turn off logging.
```
Comment (hawkw) [inline tokio-trace/src/level_filters.rs]: Nit: this wording seems a little awkward, maybe rephrase it as something like 
```suggestion
    /// Designates that logging should be to turned off.
```

File: tokio-trace/src/level_filters.rs [hunk 10]
```diff
@@ -0,0 +1,95 @@
+use std::cmp::Ordering;
+use tokio_trace_core::Level;
+
+/// Describes the level of verbosity of a span or event.
```
Comment (hawkw) [inline tokio-trace/src/level_filters.rs]: Is this comment copied from `Level`? I think it should probably describe the difference between a `Level` and a `LevelFilter`...

File: tokio-trace/src/level_filters.rs (hunk 8 above)
Comment (phungleson) [reply tokio-trace/src/level_filters.rs]: It is alright mate, thanks!

**Comment by phungleson** [comment]
Hey @hawkw thanks so much for your guidance so far. Even though I would like to add the test, and a vaguely understand the test in `log`, it seems hard for me to digest what need to be done in order to test this feature correctly. It would be great if you drops some hints in my `main.rs` copy from `log`.

Thanks again!

File: tokio-trace/test_static_max_level_features/main.rs [hunk 11]
```diff
@@ -0,0 +1,75 @@
+#[macro_use]
+extern crate log;
```
Comment (hawkw) [inline tokio-trace/test_static_max_level_features/main.rs]: We're testing tokio-trace, not `log`, so we'll want to replace this with 
```suggestion
extern crate tokio_trace;
```

File: tokio-trace/test_static_max_level_features/main.rs [hunk 12]
```diff
@@ -0,0 +1,75 @@
...
+
+use std::sync::{Arc, Mutex};
+use log::{Level, LevelFilter, Log, Record, Metadata};
+
+#[cfg(feature = "std")]
+use log::set_boxed_logger;
+#[cfg(not(feature = "std"))]
+fn set_boxed_logger(logger: Box<Log>) -> Result<(), log::SetLoggerError> {
```
Comment (hawkw) [inline tokio-trace/test_static_max_level_features/main.rs]: This function (and the import above) are specific to the `log` crate and can be removed.

File: tokio-trace/test_static_max_level_features/main.rs [hunk 13]
```diff
@@ -0,0 +1,75 @@
+#[macro_use]
+extern crate log;
+
+use std::sync::{Arc, Mutex};
+use log::{Level, LevelFilter, Log, Record, Metadata};
```
Comment (hawkw) [inline tokio-trace/test_static_max_level_features/main.rs]: we'll want to change this to import types from `tokio-trace`:
```suggestion
use tokio_trace::{field, span, Level, LevelFilter, Event, Metadata. Subscriber};
```

File: tokio-trace/test_static_max_level_features/main.rs [hunk 14]
```diff
@@ -0,0 +1,75 @@
...
+    unsafe {
+        log::set_logger(&*Box::into_raw(logger))
+    }
+}
+
+struct State {
+    last_log: Mutex<Option<Level>>,
+}
+
+struct Logger(Arc<State>);
```
Comment (hawkw) [inline tokio-trace/test_static_max_level_features/main.rs]: Since this will become a `tokio-trace` `Subscriber` instead of a `Logger`, we'll probably want to rename it to something like 
```suggestion
struct TestSubscriber(Arc<State>);
```

File: tokio-trace/test_static_max_level_features/main.rs [hunk 15]
```diff
@@ -0,0 +1,75 @@
...
+use log::set_boxed_logger;
+#[cfg(not(feature = "std"))]
+fn set_boxed_logger(logger: Box<Log>) -> Result<(), log::SetLoggerError> {
+    unsafe {
+        log::set_logger(&*Box::into_raw(logger))
+    }
+}
+
+struct State {
+    last_log: Mutex<Option<Level>>,
```
Comment (hawkw) [inline tokio-trace/test_static_max_level_features/main.rs]: I'd call this `last_level` instead of `last_log`, since we'll be testing the filtering for `tokio-trace`'s `Span` and `Event` types rather than for log messages.
```suggestion
    last_level: Mutex<Option<Level>>,
```

File: tokio-trace/test_static_max_level_features/main.rs [hunk 16]
```diff
@@ -0,0 +1,75 @@
...
+}
+
+struct Logger(Arc<State>);
+
+impl Log for Logger {
```
Comment (hawkw) [inline tokio-trace/test_static_max_level_features/main.rs]: Instead of implementing the `Log` trait from `Log`, we'll want to implement [the `Subscriber` trait](https://docs.rs/tokio-trace-core/0.1.0/tokio_trace_core/subscriber/trait.Subscriber.html) from `tokio-trace`.

Our test will only care about the [`enabled`](https://docs.rs/tokio-trace-core/0.1.0/tokio_trace_core/subscriber/trait.Subscriber.html#tymethod.enabled) and [`event`](https://docs.rs/tokio-trace-core/0.1.0/tokio_trace_core/subscriber/trait.Subscriber.html#tymethod.event) methods, which can function similarly to the `enabled` and `log` methods on the `Log` implemented here. 

We will also have to implement `record`, `record_follows_from`, `enter`, and `exit`, but (like the `flush` method on the `Log` implemented here) those can do nothing. 

For the `new_span` method, we will have to return a non-zero span ID, but we can just pick an arbitrary number and always return it (i.e. `span::Id::from_u64(42)`). When `new_span` is called, we should set the value of the `last_level` to be equal to the new span's level, similarly to what we do on `Event`s.

File: tokio-trace/test_static_max_level_features/main.rs [hunk 17]
```diff
@@ -0,0 +1,75 @@
...
+    test(&a, LevelFilter::Off);
+    test(&a, LevelFilter::Error);
+    test(&a, LevelFilter::Warn);
+    test(&a, LevelFilter::Info);
+    test(&a, LevelFilter::Debug);
+    test(&a, LevelFilter::Trace);
+}
+
+fn test(a: &State, filter: LevelFilter) {
+    log::set_max_level(filter);
```
Comment (hawkw) [inline tokio-trace/test_static_max_level_features/main.rs]: `tokio-trace` doesn't have a concept of a global max level like `log` does, so we'll want to remove this, and just try logging each message a single time. This means that we can probably get rid of the `test` function and just put the
```rust
    error!("");
    last(&a, Some(Level::ERROR);
    warn!("");
    last(&a, Some(Level::WARN);
    info!("");
    last(&a, Some(Level::INFO);

    // ...
```
bit directly in `main` instead.

File: tokio-trace/test_static_max_level_features/main.rs [hunk 18]
```diff
@@ -0,0 +1,75 @@
...
+
+    debug!("");
+    if cfg!(debug_assertions) {
+        last(&a, t(Level::Debug, filter));
+    } else {
+        last(&a, None);
+    }
+
+    trace!("");
+    last(&a, None);
```
Comment (hawkw) [inline tokio-trace/test_static_max_level_features/main.rs]: it would be nice to also have a 
```rust
span!(level: Level::TRACE, "");
last(&a, None);
```
here.

File: tokio-trace/test_static_max_level_features/main.rs [hunk 19]
```diff
@@ -0,0 +1,75 @@
...
+}
+
+fn test(a: &State, filter: LevelFilter) {
+    log::set_max_level(filter);
+    error!("");
+    last(&a, t(Level::Error, filter));
+    warn!("");
+    last(&a, t(Level::Warn, filter));
+    info!("");
+    last(&a, t(Level::Info, filter));
```
Comment (hawkw) [inline tokio-trace/test_static_max_level_features/main.rs]: let's also test that spans at enabled levels are also enabled, like
```rust
span!(level: Level::INFO, "");
last(&a, Some(Level::INFO));
```

File: tokio-trace/test_static_max_level_features/main.rs [hunk 20]
```diff
@@ -0,0 +1,75 @@
...
+        *self.0.last_log.lock().unwrap() = Some(record.level());
+    }
+
+    fn flush(&self) {}
+}
+
+fn main() {
+    let me = Arc::new(State { last_log: Mutex::new(None) });
+    let a = me.clone();
+    set_boxed_logger(Box::new(Logger(me))).unwrap();
```
Comment (hawkw) [inline tokio-trace/test_static_max_level_features/main.rs]: `tokio-trace` doesn't have a concept of a global logger. Instead, it has _scoped_ `Subscriber` contexts. So instead of this, we'll want to say something like 
```rust
    tokio_trace::subscriber::with_default(TestSubscriber(me), || {
        // run the rest of the test here...
    });
```

**Comment by phungleson** [comment]
Hey @hawkw I have updated to the best I can understand, but it is probably missing something.

Let me know if you want to add or remove anything.

File: tokio-trace/test_static_max_level_features/main.rs [hunk 21]
```diff
@@ -0,0 +1,70 @@
...
+    fn event(&self, event: &Event) {
+        *self.0.last_level.lock().unwrap() = Some(event.metadata().level().clone());
+    }
+
+    fn enter(&self, _span: &Id) {}
+
+    fn exit(&self, _span: &Id) {}
+}
+
+fn main() {
```
Comment (hawkw) [inline tokio-trace/test_static_max_level_features/main.rs]: Can we change this from a main method to a test? so
```rust
#[test]
fn test_static_max_level_features() {
// ...
}
```
I think in order for `cargo test` to pick this up, we'd also want to move the file into a`tokio-trace/test_static_max_level_features/tests/` directory.

File: tokio-trace/test_static_max_level_features/Cargo.toml [hunk 22]
```diff
@@ -0,0 +1,12 @@
+[package]
+name = "test_cargo_max_level_features"
+version = "0.1.0"
+publish = false
+
+[[bin]]
```
Comment (hawkw) [inline tokio-trace/test_static_max_level_features/Cargo.toml]: Since the test was moved into `tests`, the whole `[[bin]]` section needs to be removed.

Doing so will fix the CI build.

File: azure-pipelines.yml [hunk 23]
```diff
@@ -47,6 +47,7 @@ jobs:
       - tokio-timer
       - tokio-trace
       - tokio-trace/tokio-trace-core
+      - tokio-trace/tokio-trace-core
```
Comment (hawkw) [inline azure-pipelines.yml]: I think this was a copy-paste mistake? I believe it needs to be
```suggestion
      - tokio-trace/test_static_max_level_features
```

**Comment by phungleson** [comment]
Thanks so much for all the guidance @hawkw I have made the changes and things seem ok!

**Comment by hawkw** [comment]
Okay, I've figured out what's up with the CI build on FreeBSD: the FreeBSD build runs tests using `cargo test --all`, while the Azure Pipelines build runs `cargo test` for each individual crate. 

What this means is that on FreeBSD, one version of `tokio-trace` is compiled as a dependency for the `test_static_max_level` test crate, and this is _also_ used as the version of `tokio-trace` that's used to run the `tokio-trace` test suite. Since the dependency on `tokio-trace` in the `test_static_max_level` crate sets the `max_level_debug` feature, this feature is also set for `tokio-trace`'s own tests. This disables anything at the trace level in those tests, causing the tests to fail.

An obvious solution is to just change the Cirrus CI config to do the same thing as the Azure Pipelines CI builds, but I'd prefer to have `cargo test --all` still work. Will look into coming up with a workaround.

**Comment by hawkw** [comment]
@phungleson if I can get you to merge https://github.com/phungleson/tokio/pull/1 into your fork, I think that will resolve the FreeBSD CI issues and we'll be able to move forward with this!

**Comment by phungleson** [comment]
Looks pretty good, thanks @hawkw I have learned a lot through the process!

**Comment by hawkw** [comment]
@phungleson Great, thanks for your contribution! I'm going to go ahead and merge this when the CI build passes.

**Comment by hawkw** [comment]
aaaand....merged! thanks again @phungleson!

**Comment by phungleson** [comment]
Cool awesome! thanks again for all your help @hawkw

Do you know any particular issues that can be suitable to look into? (and hope I can be more helpful this time 😄)

File: tokio-trace/src/level_filters.rs (hunk 7 above)
Comment (phungleson) [reply tokio-trace/src/level_filters.rs]: This must be something new, I can `impl` for almost all kind of things :/

**Comment by hawkw** [comment]
@phungleson you'll probably want to keep your eye out for [issues tagged as "easy"](https://github.com/tokio-rs/tokio/issues?q=is%3Aissue+is%3Aopen+label%3Aeasy). There is also a ["good first issue" tag](https://github.com/tokio-rs/tokio-trace-nursery/issues?q=is%3Aopen+is%3Aissue+label%3A%22good+first+issue%22) in the `tokio-trace-nursery` repository, which has less stable crates to use with `tokio-trace` (although that tag is _currently_ empty).
